### ISSUE TITLE
How safe is the cached data?

### ISSUE THREAD

**Issue by MySidesTheyAreGone** [lukechilds/keyv#15]
Sorry for the awkward title.

If I `get` something and then manipulate it and the storage is a simple in-memory Map, could the manipulation (e.g. changing an object's property) affect the cached value?

Thanks!

**Comment by lukechilds** [comment]
@MySidesTheyAreGone Nope, you get a new object each time you `keyv.get()`, so it's impossible to mutate any data in the cache by manipulating an object from `keyv.get()`.

**Comment by MySidesTheyAreGone** [comment]
That's great, thanks. I would have cloned the object myself otherwise, but this way it's better.

If I may add one last question: I have a use case where a Keyv instance backed by a store that doesn't support TTL functionality (memory, sqlite) would grow indefinitely (because keys aren't deleted proactively unless, you know, one uses Redis or something like that). That's because data would make it into the cache once but would never be `get`ted again, leaving the memory / disk space occupied.

I understand that iterating over the entire store on Keyv's side would impact all of the namespaces, so some filtering of the keys would be necessary. Doing this filtering in a store-independent way would be probably impossible. 

Moreover, there's no way of avoiding `get`ting each and every cached item in the current namespace in order to look at the expiration time; so the only reasonable way of doing a sweep would be to have Keyv also keep a chronologically sorted list of keys and expiration dates that could be used to cut down on the performance impact at the expense of a bit more memory. This part could be disabled by default and enabled on demand, but still, it's quite a bit of code.

Given all this, is a `cleanup` function of sorts worth figuring out and including in Keyv? If you're open to the idea I can work on a pull request, but this is a magnificently simple module and I'd understand if you'd rather not impact it with something so intrusive. The alternative - the user bites the bullet and just calls `get` on every key periodically - is not unacceptable, after all. So before acting, I thought I'd just ask for your opinion.

**Comment by lukechilds** [comment]
Interesting point, so yeah my absolute number one goal is to keep this as simple as possible. It simulates the [Map API](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map) so it's unlikely that I'll add any features that aren't already available on `new Map()`. There is already a `.clear()` method but that just nukes the current namespace, regardless of TTL.

My main stance on the issue is that the way TTLs function on the application side with Keyv on either SQLite or Redis are identical, which is the problem Keyv is trying to solve. If you need to expire your TTLs instantly on the DB side then you should back Keyv with a DB that has this functionality built in.

**Comment by lukechilds** [comment]
@MySidesTheyAreGone Also, just FYI, you can solve this problem with an in-memory adapter by using a third party LRU memory store that follows the Map API like `quick-lru`

See example in the readme: https://github.com/lukechilds/keyv#third-party-storage-adapters

Still an issue with SQL DBs though obvs

**Comment by MySidesTheyAreGone** [comment]
Wait you're right, I think I get it now: the TTL functionality belongs on the store side, not in Keyv. It's obvious that Keyv is solving a completely different problem. My question looks very foolish now lol

`quick-lru` is a perfect example. And if I go, say, for sql storage I'm sure I can easily extend keyv-sql (I just looked at it) or one of the others so that it sweeps away old stuff. Thanks for everything!

**Comment by lukechilds** [comment]
Yeah, you could fork keyv-sql and update the schema to have a timestamp column and periodically clear entries with expired timestamps. You could then just fork keyv-sqlite/postgres/mysql and just change them to extend your keyv-sql fork it would all just magically work.

This isn't something that I'm interested in merging into core but would definitely solve your use case and could be third party storage adapters that others may find useful.

**Comment by MySidesTheyAreGone** [comment]
Hey let me know if there are any problems with livensing/attribution/logos/whatever. Stuff should be OK and working, I'm doing some real world testing.

https://github.com/MySidesTheyAreGone/keyv-sql-shrink

https://github.com/MySidesTheyAreGone/keyv-sqlite-shrink

Thanks again

**Comment by lukechilds** [comment]
All good, nice work 👍 

The title in the https://github.com/MySidesTheyAreGone/keyv-sqlite-shrink readme is still `@keyv/sqlite` btw.
